pkg/route/api/validation/validation.go

```diff
@@ -102,9 +108,121 @@ func validateTLS(tls *routeapi.TLSConfig) fielderrors.ValidationErrorList {
 		if len(tls.DestinationCACertificate) > 0 {
 			result = append(result, fielderrors.NewFieldInvalid("destinationCACertificate", tls.DestinationCACertificate, "edge termination does not support destination certificates"))
 		}
+
+		result = append(result, validateTLSCertificates(tls)...)
+
 	default:
 		msg := fmt.Sprintf("invalid value for termination, acceptable values are %s, %s, %s, or emtpy (no tls specified)", routeapi.TLSTerminationEdge, routeapi.TLSTerminationPassthrough, routeapi.TLSTerminationReencrypt)
 		result = append(result, fielderrors.NewFieldInvalid("termination", tls.Termination, msg))
 	}
 	return result
 }
+
+// validateTLSCertificates checks that the certificates passed are able to be parsed.  This includes ensuring that
+// the cert/key pair are matching and, if a ca cert is provided it will check the first ca cert against the leaf
+// to ensure that the leaf was signed by the ca
+func validateTLSCertificates(tls *routeapi.TLSConfig) fielderrors.ValidationErrorList {
+	if tls == nil {
+		return nil
+	}
+	//edge and reencrypt (the use cases that call this require all 3 (at least) so we can ignore anything
+	//that doesn't have all 3 set.  It will already fail validation elsewhere
+	if len(tls.Certificate) == 0 || len(tls.Key) == 0 || len(tls.CACertificate) == 0 {
+		return nil
+	}
+
+	result := fielderrors.ValidationErrorList{}
+
+	//check cert/key pair
+	checkSignature := true
+	cert, err := ctls.X509KeyPair([]byte(tls.Certificate), []byte(tls.Key))
```

Comment:
Ok, that reduces some of my concern.  We may need to handle PKCS#7 and #12 but for now we could require DER.  We would need to make sure it's possible in a router implementation to convert formats eventually (i.e. the router can do that somehow).

----- Original Message -----
> > +// to ensure that the leaf was signed by the ca
> > +func validateTLSCertificates(tls *routeapi.TLSConfig)
> > fielderrors.ValidationErrorList {
> > +	if tls == nil {
> > +		return nil
> > +	}
> > +	//edge and reencrypt (the use cases that call this require all 3 (at
> > least) so we can ignore anything
> > +	//that doesn't have all 3 set.  It will already fail validation elsewhere
> > +	if len(tls.Certificate) == 0 || len(tls.Key) == 0 ||
> > len(tls.CACertificate) == 0 {
> > +		return nil
> > +	}
> > +
> > +	result := fielderrors.ValidationErrorList{}
> > +
> > +	//check cert/key pair
> > +	checkSignature := true
> > +	cert, err := ctls.X509KeyPair([]byte(tls.Certificate), []byte(tls.Key))
> 
> Some references.  It looks like the big players I'd expect to see as a router
> all require PEM format
> 
> haproxy -
> http://cbonte.github.io/haproxy-dconv/configuration-1.5.html#5.1-crt
> nginx -
> http://nginx.org/en/docs/http/ngx_http_ssl_module.html#ssl_certificate
> apache - http://httpd.apache.org/docs/2.4/mod/mod_ssl.html#sslcertificatefile
> f5 - https://support.f5.com/kb/en-us/solutions/public/6000/500/sol6549.html
> 
> More details from my understanding:
> 
> DER/PEM: these are encoding formats of x509 certs.  DER is a binary encoded
> version of an ASCII PEM.   DER can be converted to PEM format with openssl
> 
> PKCS: a packaging format for cert distribution.  PKCS#7 can be used to bundle
> certs and cannot contain a private key.  PKCS#12 is used for packaging cert
> chains and can contain the private key.  Major use of PKCS is IIS and java
> key stores (tomcat).  openssl can convert pkcs certificates to PEM/DER
> format
> 
> 
> /cc @rajatchopra  - Rajat, is this too restrictive for validation?  Are there
> common router implementations that I'm not considering that would fail in
> golang's parsing of a PEM encoded x509 key pair?
> 
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/openshift/origin/pull/1824/files#r28778758